### ISSUE TITLE
Team issues

### ISSUE THREAD

**Issue by mkalininait** [rails-girls-summer-of-code/rgsoc-teams#430]
+ [ ] We restrict adding a coach, who is already a student in another team. At the same time I can add a student who is already a could in another team.
+ [ ] When I try to add more than 2 students: "Roles there cannot be more than 2 students on a team." <- weird phrase. 
+ [ ] There is also the message "Roles user has already been taken", and I also don't know what it means.
+ [ ] Managed to get 2 error boxes when adding a new team member. This is not user friendly. Can we show only one box?

![screen shot 2016-03-16 at 23 44 16](https://cloud.githubusercontent.com/assets/1745735/13831070/317de83c-ebd1-11e5-9d94-ba6fb276cd2c.png)

+ [ ] When I edit a team, I can edit roles of team members. Is it a bug or a feature?
Here is a possible scenario: I added a member and set a wrong role for him/her (say coach instead of student). I press 'Edit', but I can't change the role, if I delete this member and add him/her again with another role, I get the error ''X can't have more than one role in this team!" (which happens because of [Issue 420](https://github.com/rails-girls-summer-of-code/rgsoc-teams/issues/420). The solution would be to delete, save and then add the member again, but it's not a good UX.

+ [ ] Can one person be a coach for more than one team? Now it's possible.

+ [ ] Little wish: make 'Help section' link open in a new tab.

**Comment by klappradla** [comment]
> This is not user friendly. Can we show only one box?

I added the second box in order to make it more user-friendly 🙊 . With only one box, you get a error-message saying:

    2 errors prohibited this role from being saved:
        User has already been taken
        Team is invalid

*Team is invalid* is not really a meaningful error-message for a user not knowing the internal datamodel of the application (anyone basically...), therefore I added what's the current approach.

I also thought about fully removing the **`Add member`** button, but then decided against, because it acts as a handy shortcut for admins to look and edit a team's members. But if it causes more trouble than benefit, we can also remove it - it does nothing the teams-form does not already do.

**Comment by mkalininait** [comment]
Tested this ticket on Staging: all unchecked problems are still relevant.

**Comment by bitboxer** [comment]
For 1: what should be done here? This is just saying what the current status is but not what should be done.
For 2, 3: what should the wording be?
For 4: is this still relevant after the comment by klappradia?
For 5: does this need fixing or is this considered a bug?

**Comment by mkalininait** [comment]
@bitboxer Hey, sorry for not replying yesterday, i got stuck on the first question.

1. We should restrict adding user as a student if he/she is already added as a coach (confirmed or not confirmed, doesn't matter) in another team. A user can always delete him/herself as a coach from a team in order to become a student (coaches have this permission). 

Btw, here is a nice example: Armin has been added to the [Team Oberlaa](https://rgsoc-teams-staging.herokuapp.com/teams/127) as a coach. Then he has been added to the [Team Sleepy Testers](https://rgsoc-teams-staging.herokuapp.com/teams/128) as a student. As a result of this, we can't edit the Team Oberlaa anymore because whatever you do you get this: "1 error prohibited this team from being saved: Armin Ronacher already is a student on another team for 2017."

**Comment by bitboxer** [comment]
Yes, I understand that this is the problem. But what do you want implemented as solution to this?

**Comment by mkalininait** [comment]
@bitboxer 
2. "Roles: there cannot be more than 2 students on a team."
3. "Roles: user has already been taken" (If we use this error message only when trying to add one user twice within **one** team, the message could be "Roles: user has already been added". But I can't tell it just by looking at the UI, it should probably be clearer after reading the code).

I found two more error messages of this type:
 * "Roles name is not included in the list"
 * "Roles name can't be blank"

I'm not sure how many of them there are in the app. Maybe all we do is just add a colon after the word *Roles* for **all** error messages: "Roles" --> "Roles:"? WDYT?

**Comment by mkalininait** [comment]
@klappradla Thanks for the comment.
In that particular screenshot, the error message in the second box repeats on of the error messages in the first box, that's why it looked redundant to me. Maybe there are other use cases though, when the second box provides more useful information. I just didn't see them...
If you vouch for 2 boxes, then we won't change it :)
//cc @bitboxer

**Comment by mkalininait** [comment]
@bitboxer 
5. I'm fine skipping it. WDYT @alicetragedy ?

**Comment by bitboxer** [comment]
For the error boxes: I will try to make it look like one big box with two messages inside.

**Comment by bitboxer** [comment]
For the message in the error text: this message is generated from Rails and uses the following convetion: "[Attribute] [Error Message]". I don't want to add a colon between the attribute and the message, this would mean we have to check every error message. I think the best approach is to form a sentence with "Role " at the beginning. For example: "Role need fixing. The user [USERNAME] has already be taken" . Would that be okay?

**Comment by bitboxer** [comment]
And for the "name can't be blank": This is sadly how rails creates the messages. 

Ah, another way would be to inline the error messages into the form and show the text next to the field that has the error. But we would need to redo every form to stay consistant. That is also lots of work. 

Maybe we can just fix the two text messages and just ignore it and hope it works for most students?

**Comment by bitboxer** [comment]
@mkalininait I tried to reproduce the two error message boxes, but I can't. It totally looks different on my machine:

<img width="1213" alt="bildschirmfoto 2017-01-23 um 14 44 48" src="https://cloud.githubusercontent.com/assets/56195/22206201/9863bdf2-e17a-11e6-9069-d78afc52f45b.png">

I assume now it is already fixed. If not, please post details how to reproduce it with a new screenshot.

With all what we have discussed here, the things left to do:

* Prevent Coaches to be Team Members in teams. This makes the "delete them" part obsolete, because you won't be able to create such teams.
* Change the error messages a bit to make them clearer.

**Comment by bitboxer** [comment]
After playing around with the coach / student rules I figured out that currently the system is setup in a way that everyone can get every role as often as they want. With the expection that a person can always be additionally added as a student in other teams. But once this person is a student somewhere, that person can't be removed or edited in other teams because the check if the person is a student prevents this.

There are now two of ways to fix this:

* Prevent a Person to be on more than one team in whatever role.
* Prevent a Person to become a student in a team when that person is already added in other teams in whatever role

So this problem is not only limited to coaches, but to everyone who is added as a student and already has a role somewhere.

What should I do? The first or the second version? Currently I think I would vote for the 2nd one because it could be okay that a person coaches more than one team. Am I right?

**Comment by mkalininait** [comment]
Hi @bitboxer 
Regarding the **"coach / student rules"**.

>  it could be okay that a person coaches more than one team. Am I right?  

That's right, one person can coach more than one team.

> The first or the second version?  

The 2nd looks correct to me. Note: we should check teams within current season.

**Comment by mkalininait** [comment]
@bitboxer 
Regarding **two error boxex**. 

I still can reproduce this bug locally. It's important that you try to add the same user with the same role. E.g. there is a Team with student X and coach Y. One box shows up if you try to add X as a coach or Y as a student in this team. But if you try to add X as a student or Y as a coach again, you'll see two boxes.

![team-error-boxes](https://cloud.githubusercontent.com/assets/1745735/22224899/b29780aa-e1c0-11e6-89e5-630c0d6f9959.gif)

**Comment by mkalininait** [comment]
@bitboxer 
Regarding the **error messages**.

Thanks for clarifying why they look this way at the moment. That's interesting.

We can definitely start the error messages with *"Roles"* (or *"Role"*, whatever the name of this attribute is). If we want to change individual messages, we could also change them in a way that they start with a colon, couldn't we? :) Anyway, I don't really have a strong opinion on how the texts should look like (**"Roles: [blah-blah]" or "Roles need fixing. [Blah-blah]" or "Roles error. [Blah-blah]"**) – as long as they are clear and grammatically correct. I'm fine with the option which you consider to be the best :raised_hands:

**Comment by carpodaster** [comment]
Hey @mkalininait do you find the time to triage your list in the issue description and see what's left? Ideally, we'll take the remainders into new, atomic issues as this one is difficult to manage due to its amount of comments
